rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,406 @@
+14. Replace the old symlinks/binaries and `/var/lib/teleport/proc/sqlite.db` and quit (exit 1) if unsuccessful.
+15. Remove all stored versions of the agent except the current version and last working version.
+
+To enable auto-updates of the updater itself, all commands will first check for an `active_version`, and reexec using the `teleport-updater` at that version if present and different.
+The `/usr/local/bin/teleport-updater` symlink will take precedence to avoid reexec in most scenarios.
+
+To ensure that SELinux permissions do not prevent the `teleport-updater` binary from installing/removing Teleport versions, the updater package will configure SELinux contexts to allow changes to all required paths.
+
+To ensure that `teleport` package removal does not interfere with `teleport-updater`, package removal will run `apt purge` (or `yum` equivalent) while ensuring that `/etc/teleport.yaml` and `/var/lib/teleport` are not purged.
+Failure to do this could result in `/etc/teleport.yaml` being removed when an operator runs `apt purge` at a later date.
+
+#### Failure Conditions
+
+If the new version of Teleport fails to start, the installation of Teleport is reverted as described above.
+
+If `teleport-updater` itself fails with an error, and an older version of `teleport-updater` is available, the upgrade will retry with the older version.
+
+Known failure conditions caused by intentional configuration (e.g., upgrades disabled) will not trigger retry logic.
+
+#### Status
+
+To retrieve known information about agent upgrades, the `status` subcommand will return the following:
+```json
+{
+  "agent_version_installed": "15.1.1",
+  "agent_version_desired": "15.1.2",
+  "agent_version_previous": "15.1.0",
+  "agent_edition_installed": "enterprise",
+  "agent_edition_desired": "enterprise",
+  "agent_edition_previous": "enterprise",
+  "agent_update_time_last": "2020-12-10T16:00:00+00:00",
+  "agent_update_time_jitter": 600,
+  "agent_updates_enabled": true
+}
+```
+
+### Downgrades
+
+Downgrades may be necessary in cases where we have rolled out a bug or security vulnerability with critical impact.
+To initiate a downgrade, `agent_version` is set to an older version than it was previously set to.
+
+Downgrades are challenging, because `sqlite.db` used by newer version of Teleport may not be valid for older versions of Teleport.
+
+When Teleport is downgraded to a previous version that has a backup of `sqlite.db` present in `/var/lib/teleport/versions/OLD-VERSION/backup/`:
+1. `/var/lib/teleport/versions/OLD-VERSION/backup/backup.yaml` is validated to determine if the backup is usable (proxy and version must match, age must be less than cert lifetime, etc.)
+2. If the backup is valid, Teleport is fully stopped, the backup is restored along with symlinks, and the downgraded version of Teleport is started.
+3. If the backup is invalid, we refuse to downgrade.
+
+Downgrades are applied with `teleport-updater update`, just like upgrades.
```

Comment:
It's a bit unclear if the local version from the `/var/lib/teleport/versions` directory is used, or if a new one is downloaded as the update flow describes.

If the former, does that include checksum verification? Technically, if the older version is stored as a tarball, the original checksum information could be used. But if the tarball is already unpacked, there are no checksums recorded, and any malicious modification or corruption will not be detected. This is still happening within the /var/lib/teleport directory, so the risk is small, but it might be worth calling out explicitly.

Reply:
The local, uncompressed version is used. This is so that the rollback uses as little logic as possible, the restored state is as similar as possible to what was running before, and so that recovery happens as quickly as possible.

We are already trusting that `/var/lib/teleport` isn't maliciously modified or corrupted when, e.g., we restart the service, but I agree this should be called out explicitly in the RFD.

Reply:
https://github.com/gravitational/teleport/pull/40190/commits/00a1ea07a53abf6a253da0503140c2035ac3dad2